compiler/src/iree/compiler/Dialect/Stream/Transforms/SpecializeEncodings.cpp

```diff
@@ -0,0 +1,172 @@
+  funcOp.walk([&](AffinityOpInterface affinityOp) {
+    // Only need to update encoding types for ops that have TensorPhaseOp trait.
+    if (!affinityOp->hasTrait<OpTrait::IREE::Stream::TensorPhaseOp>()) {
+      return;
+    }
+
+    // Bail out if the operation does not have an affinity attribute.
+    // TODO(hanchung): We should use the default device in this case. However,
+    // it is not guaranteed that default device attribute will always be set in
+    // the IR. (Is the statement correct?)
+    auto affAttr = affinityOp.getAffinityAttr();
+    if (!affAttr) {
+      return;
+    }
+    candidates.push_back(affinityOp);
+  });
+
+  if (candidates.empty()) {
+    return success();
+  }
+
+  IRRewriter rewriter(funcOp.getContext());
+  for (auto affinityOp : candidates) {
+    auto affAttr = affinityOp.getAffinityAttr();
+    SetVector<Attribute> layouts;
+    if (failed(makeLayoutAttrFn(affAttr, moduleOp, layouts))) {
+      affinityOp.emitError("failed on making layouts");
+      return failure();
+    }
+
+    auto getEncodingWithNewLayouts =
+        [=](Type type) -> std::optional<IREE::Encoding::EncodingAttr> {
+      auto rankedTensorType = dyn_cast<RankedTensorType>(type);
+      if (!rankedTensorType) {
+        return std::nullopt;
+      }
+      auto encoding = IREE::Encoding::getEncodingAttr(rankedTensorType);
+      if (!encoding) {
+        return std::nullopt;
+      }
+      SmallVector<Attribute> attrs(layouts.begin(), layouts.end());
+      return encoding.cloneWithLayouts(attrs);
+    };
+    // TODO(hanchung): Update other Stream operations.
+    LogicalResult result =
+        TypeSwitch<Operation *, LogicalResult>(affinityOp)
+            .Case<Stream::TensorSizeOfOp>([&](auto sizeOfOp) {
+              auto encodingType =
+                  dyn_cast<RankedTensorType>(sizeOfOp.getEncoding());
+              if (!encodingType) {
+                return success();
+              }
+              std::optional<IREE::Encoding::EncodingAttr> encoding =
+                  getEncodingWithNewLayouts(encodingType);
+              if (!encoding.has_value()) {
+                return success();
+              }
+              rewriter.modifyOpInPlace(sizeOfOp, [&] {
+                sizeOfOp.setEncoding(
+                    cloneWithEncoding(encodingType, encoding.value()));
+              });
+              return success();
+            })
+            .Default([](auto *op) { return success(); });
+
+    if (failed(result)) {
+      return failure();
+    }
+  }
+  return success();
+}
+} // namespace
+
+struct SpecializeEncodingsPass
+    : public impl::SpecializeEncodingsPassBase<SpecializeEncodingsPass> {
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+    auto usedDialects =
+        gatherUsedDialectInterfaces<AffinityAnalysisDialectInterface>(moduleOp);
+    if (usedDialects.size() != 1) {
+      moduleOp.emitError("expected only one dialect implementing "
```

Comment:
we don't want this check - if it's required to make the rest work we should make the rest work - we want to allow other dialects (our own and others) to implement the interface. to start you could just loop over the dialects inside the funcOp loop below (so for each func op apply layouts for each dialect).

Reply:
I'm being careful here because the interface is new. It is not clear to me what the other use cases would look like.

I don't follow what you meant by looping it over the dialects in side the funcOp loop. Is it something like below?

If so, there is a completeness issue to me. E.g., it seems like we should append layouts to encodings while the current implementation is replacing the layouts in encodings. The other issue (which is not present in this PR) is that I don't know how to interpret it with the executable duplication and updating dispatch. They are HAL specific and I image that other dialect implementation would only apply it on their own dialect ops. It looks like I need to add additional checks (that I do not know what they are) to the implementation. It adds potential bugs and breaks my mental model to the pass. In this context, I'd prefer bailing out in the first place. People can come and add such support when they have needs. I can also help on it after I have clearer picture. What do you think?

```cpp
for (auto funcOp : moduleOp.getOps<mlir::FunctionOpInterface>()) {
  for (dialectLayoutAttrResolver: usedDialects) {
    ... addLayoutsToTensorPhaseOps(moduleOp, funcOp, dialectLayoutAttrResolver) ...
  }
}
```